File: h/树的子结构.java
/*
输入两棵二叉树A，B，判断B是不是A的子结构。
（ps：我们约定空树不是任意一个树的子结构）
*/
/**
public class TreeNode {
    int val = 0;
    TreeNode left = null;
    TreeNode right = null;

    public TreeNode(int val) {
        this.val = val;

    }

}
*/
public class Solution {
    public boolean HasSubtree(TreeNode pRoot1, TreeNode pRoot2) {

         boolean flag = false;
         if ((pRoot1 != null) && (pRoot2 != null)) {
             if (pRoot1.val == pRoot2.val) {
                flag = HasSubtreetty(pRoot1, pRoot2);
             }
     
             if (!flag) {
                flag = HasSubtree(pRoot1.left, pRoot2);
             }
 
             if (!flag) {
                flag = HasSubtree(pRoot1.right, pRoot2);
             }   
         }
         return flag;
    }
    
    public boolean HasSubtreetty(TreeNode pRoot1, TreeNode pRoot2) {

        if (pRoot2 == null) {
            return true;
         }
            
        if ((pRoot1 == null) && (pRoot2 != null)) {
            return false;
        }
           
        if (pRoot1.val != pRoot2.val) {
            return false;
        }
     
         return HasSubtreetty(pRoot1.left, pRoot2.left) && HasSubtreetty(pRoot1.right, pRoot2.right);
     }
}

File: z/树的子结构.cc
/*
struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
    TreeNode(int x) :
            val(x), left(NULL), right(NULL) {
    }
};*/
class Solution {
 public:
  bool helper(TreeNode* node1, TreeNode* node2) {
    if (node2 == NULL) {
      return true;
    }
    if (node1 == NULL) {
      return false;
    }
    if (node1->val != node2->val) {
      return false;
    }
    return helper(node1->left, node2->left) && helper(node1->right, node2->right);
  }
  bool HasSubtree(TreeNode* pRoot1, TreeNode* pRoot2) {
    if (pRoot2 == NULL) {
      return false;
    }
    bool res = false;
    if (pRoot1 != NULL && pRoot2 != NULL) {
      if (pRoot1->val == pRoot2->val) {
        res = helper(pRoot1, pRoot2);
      }
      if (res == false) {
        res = HasSubtree(pRoot1->left, pRoot2);
      }
      if (res == false) {
        res = HasSubtree(pRoot1->right, pRoot2);
      }
    }
    return res;
  }
};
